pkg/option/config.go

```diff
@@ -2735,6 +2711,160 @@ func (c *DaemonConfig) populateHostServicesProtos() error {
+			c.PolicyMapEntries, PolicyMapMin)
+	}
+	if c.PolicyMapEntries > PolicyMapMax {
+		return fmt.Errorf("specified PolicyMap max entries %d must not exceed maximum %d",
+			c.PolicyMapEntries, PolicyMapMax)
+	}
+
+	if c.FragmentsMapEntries < FragmentsMapMin {
+		return fmt.Errorf("specified fragments tracking map max entries %d must exceed minimum %d",
+			c.FragmentsMapEntries, FragmentsMapMin)
+	}
+	if c.FragmentsMapEntries > FragmentsMapMax {
+		return fmt.Errorf("specified fragments tracking map max entries %d must not exceed maximum %d",
+			c.FragmentsMapEntries, FragmentsMapMax)
+	}
+
+	return nil
+}
+
+func (c *DaemonConfig) calculateBPFMapSizes() error {
+	// BPF map size options
+	// Any map size explicitly set via option will override the dynamic
+	// sizing.
+	c.CTMapEntriesGlobalTCP = viper.GetInt(CTMapEntriesGlobalTCPName)
+	c.CTMapEntriesGlobalAny = viper.GetInt(CTMapEntriesGlobalAnyName)
+	c.NATMapEntriesGlobal = viper.GetInt(NATMapEntriesGlobalName)
+	c.PolicyMapEntries = viper.GetInt(PolicyMapEntriesName)
+
+	// Allow the range (0.0, 1.0] because the dynamic size will anyway be
+	// clamped to the table limits. Thus, a ratio of e.g. 0.98 will not lead
+	// to 98% of the total memory being allocated for BPF maps.
+	dynamicSizeRatio := viper.GetFloat64(MapEntriesGlobalDynamicSizeRatioName)
+	if 0.0 < dynamicSizeRatio && dynamicSizeRatio <= 1.0 {
+		vms, err := mem.VirtualMemory()
+		if err != nil || vms == nil {
+			log.WithError(err).Fatal("Failed to get system memory")
+		}
+		c.calculateDynamicBPFMapSizes(vms.Total, dynamicSizeRatio)
+	} else if dynamicSizeRatio < 0.0 {
+		return fmt.Errorf("specified dynamic map size ratio %f must be ≥ 0.0", dynamicSizeRatio)
+	}
+	return nil
+}
+
+func (c *DaemonConfig) calculateDynamicBPFMapSizes(totalMemory uint64, dynamicSizeRatio float64) {
+	// Heuristic:
+	// Distribute relative to map default entries among the different maps.
+	// Cap each map size by the maximum. Map size provided by the user will
+	// override the calculated value and also the max. There will be a check
+	// for maximum size later on in DaemonConfig.Validate()
+	//
+	// Calculation examples:
+	//
+	// Memory   CT TCP  CT Any      NAT  Policy
+	//
+	//  512MB    33140   16570    33140    1035
+	//    1GB    66280   33140    66280    2071
+	//    4GB   265121  132560   265121    8285
+	//   16GB  1060485  530242  1060485   33140
+	memoryAvailableForMaps := int(float64(totalMemory) * dynamicSizeRatio)
+	log.Debugf("Memory available for map entries (%.3f of %d): %d", dynamicSizeRatio, totalMemory, memoryAvailableForMaps)
+	//                      Key     Value   Entry
+	//  CT(tcp + any) IPv4: 14    + 56    = 70
+	//                IPv6: 38    + 56    = 94
+	//  NAT IPv4:           14    + 38    = 52
+	//      IPv6:           38    + 50    = 88
+	//  Policy:             8     + 24    = 32
+	sizeofCTEntry := 94
+	sizeofNATEntry := 88
+	sizeofPolicyEntry := 32
```

Comment:
Any chance we could compute those from the sizes of the actual structs?

Reply:
Unfortunately, these structs are defined in `pkg/maps` which in turn depends on `pkg/option` which would lead to a circular dependency, which I'd like to avoid f possible. I agree that specifying these sizes manually is far from optimal but then again the point is not to get them exactly right, but to have them just approximately right so the available memory will be distributed fairly.

Reply:
I thought we might hit a circular dependency but wasn't sure. Ok, no problem, thanks for looking into it!

Reply:
@tklauser we can add these values as fields part of the `DaemonConfig` and set them before calling `Populate`, here
https://github.com/cilium/cilium/blob/93b209b9342fb2901b95358f6ad11f176cfaed70/daemon/cmd/daemon_main.go#L799-L801

Reply:
Thanks, this turned up that I was actually underestimating the NAT map entry size!